### ISSUE TITLE
TS types for pages

### ISSUE THREAD

**Issue by jfsiii** [frandiox/vitedge#51]
Currently, (as far as I can tell), pages do not receive typed props

e.g.  [`examples/react/src/pages/Home.jsx`](https://github1s.com/frandiox/vitedge/blob/master/examples/react/src/pages/Home.jsx#L4)
<img width="532" alt="Screen Shot 2021-08-15 at 7 28 03 PM" src="https://user-images.githubusercontent.com/57655/129495859-89d7c32c-a993-4dca-a7e4-28a6428703e5.png">

Ideally, a page would get the types automatically. Worst case, a page could import a type (from `functions/{props,api}` or elsewhere) but I believe it's really important the pages receive typed props.

Perhaps this is all under the "Rewrite in TypeScript" TODO, but I wanted to highlight this specific case and ask for it to be prioritized.

**Comment by jfsiii** [comment]
Let me know if I'm missed something and it is possible today. I'm happy to help, especially if you've already got an idea of how/where to do it. 

It seems like step 1 is making the connection between the return type of `EdgeProps['handler']` and a page, but even that still leaves us with `any` with the current types

<img width="511" alt="Screen Shot 2021-08-15 at 7 50 15 PM" src="https://user-images.githubusercontent.com/57655/129496386-acf215de-c947-4d2b-b6fe-e8e904ae3db2.png">

I'm not sure if we can get `Home` to have the exact type returned by its specific `propGetter` or if it would have to

```ts
interface HomeProps {
  server: boolean;
  msg: string;
}
```

**Comment by frandiox** [comment]
@jfsiii Hi!
The thing is, the props for each page comes from user-land since you can basically return anything you need from the props handler (that's why the `data` property is `any`).
Therefore, the only thing that comes to mind is having a place where you export your own types. Then, import the types in both your props-handler and your page.

I'm not that fluent in TS-compiler stuff so not sure if this can be automated in any way 🤔  Do you have any proposal?

**Comment by jfsiii** [comment]
> The thing is, the props for each page comes from user-land since you can basically return anything you need from the props handler (that's why the data property is any).

@frandiox  Right. But we can create a type for that response, or let the user supply it. There are a few options depending on the DX we want, but one option is to export the type `handler` returns and let it accept a Type (i.e. a Generic). e.g. with

```ts
// index.d.ts
export type ReturnedPropsPayload<T=any> = PropsOptions & { data?: T }
```

it can be used like
```diff
diff --git a/examples/react/functions/props/post.ts b/examples/react/functions/props/post.ts
index d664954..a2229bd 100644
--- a/examples/react/functions/props/post.ts
+++ b/examples/react/functions/props/post.ts
@@ -1,13 +1,21 @@
+import type { ReturnedPropsPayload } from 'vitedge';
 import { defineEdgeProps } from 'vitedge/define'

+export interface PropsData {
+  server: boolean,
+  params: Record<string, unknown>,
+}
+
+type EdgePropsData = ReturnedPropsPayload<PropsData>
+
 export default defineEdgeProps({
-  async handler({ params = {}, query = {} }) {
+  async handler({ params = {} }): Promise<EdgePropsData> {
     return {
       data: {
         server: true,
         params,
       },
    }
    };
   },
   options: {
     cache: {
```

<details><summary>and be type-safe (screenshots from VS Code)</summary>

<img width="909" alt="Screen Shot 2021-08-23 at 7 32 57 PM" src="https://user-images.githubusercontent.com/57655/130533606-c0e5a92d-bba5-4b11-8fb1-30835bb47214.png">
<img width="845" alt="Screen Shot 2021-08-23 at 7 33 10 PM" src="https://user-images.githubusercontent.com/57655/130533607-7db9fa42-343d-4f28-8780-352fd5903815.png">
<img width="952" alt="Screen Shot 2021-08-23 at 7 33 28 PM" src="https://user-images.githubusercontent.com/57655/130533609-5f9730fc-d43d-4037-aa72-bcba89c5bc57.png">
<img width="596" alt="Screen Shot 2021-08-23 at 7 35 37 PM" src="https://user-images.githubusercontent.com/57655/130533611-5766a814-be6b-4a92-9837-d466e5939ff2.png">
<img width="889" alt="Screen Shot 2021-08-23 at 7 35 49 PM" src="https://user-images.githubusercontent.com/57655/130533613-f95e19c9-7bb4-4989-b7bb-162ec4bf1e85.png">
<img width="740" alt="Screen Shot 2021-08-23 at 7 35 58 PM" src="https://user-images.githubusercontent.com/57655/130533614-d621e62a-7fb3-44aa-9b8a-fa03ceeb5096.png">
<img width="585" alt="Screen Shot 2021-08-23 at 7 36 11 PM" src="https://user-images.githubusercontent.com/57655/130533618-071b696a-04e4-43aa-8c65-55b7804a6a9a.png">

</details>

That at least gives a typed props handler. There are similar options with the API endpoints and page handlers. 

Ideally, vitedge would make the connections so the page handlers would know specifically which props they received (Home, About, etc) but, worst case, the user could take a similar approach as before

```diff
diff --git a/examples/react/src/pages/post/index.jsx b/examples/react/src/pages/post/index.jsx
index b63ca64..c0bb013 100644
--- a/examples/react/src/pages/post/index.jsx
+++ b/examples/react/src/pages/post/index.jsx
@@ -1,6 +1,14 @@
 import React from 'react'
+// type we defined/exported earlier
+import type { PropsData } from '../../../functions/props/post';

-const Post = ({ params = {}, isLoadingProps, isRevalidatingProps }) => {
+// comes from vitedge or somewhere outside page
+type PageProps = {
+  isLoadingProps: boolean
+  isRevalidatingProps: boolean
+};
+
+const Post = (props: PageProps & PropsData) => {
+  const { params, isLoadingProps, isRevalidatingProps, server,  } = props;
   if (!import.meta.env.SSR) {
     console.log({ isLoadingProps, isRevalidatingProps })
   }
```

<details><summary>Screenshots for page handler</summary>

<img width="898" alt="Screen Shot 2021-08-23 at 8 12 32 PM" src="https://user-images.githubusercontent.com/57655/130536623-8ca119df-60aa-4816-b28f-7e3ae64fd8ea.png">
<img width="708" alt="Screen Shot 2021-08-23 at 8 26 29 PM" src="https://user-images.githubusercontent.com/57655/130536624-3a45156f-fc56-46be-b349-6c4584808c47.png">

Unfortunately, this is `params` is still `Record<string, unknown>` vs something route-specific like `{ postId: string }`
</details>

**TL;DR;** I'm sure we can improve things; it's largely a matter of taste (desired DX) and some work defining the types

**Comment by jfsiii** [comment]
Also, if we define the routes a bit differently

<details><summary>diff showing individual routes with unique `name` and `path`</summary>

```diff
diff --git a/examples/react/src/routes.js b/examples/react/src/routes.js
index 81695be..d98e813 100644
--- a/examples/react/src/routes.js
+++ b/examples/react/src/routes.js
@@ -1,30 +1,43 @@
 import { createElement } from 'react'
 
-export default [
-  {
-    path: '/',
-    name: 'home',
-    exact: true,
-    component: () => import('./pages/Home'),
-  },
-  {
-    path: '/about',
-    name: 'about',
-    component: () => import('./pages/About'),
-  },
-  {
-    path: '/post/:postId',
-    name: 'post',
-    component: () => import('./pages/post'),
-    meta: {
-      propsGetter: 'post',
-    },
+export const HomeRoute = {
+  path: '/' as const,
+  name: 'home' as const,
+  exact: true as const,
+  component: () => import('./pages/Home'),
+};
+
+export const AboutRoute = {
+  path: '/about' as const,
+  name: 'about' as const,
+  component: () => import('./pages/About'),
+};
+
+export const GetPostByIdRoute = {
+  path: '/post/:postId' as const,
+  name: 'post' as const,
+  component: () => import('./pages/post'),
+  meta: {
+    propsGetter: 'post' as const,
   },
-].map(({ component: fn, ...route }) => {
+};
+
+
+export default [
+  HomeRoute,
+  AboutRoute,
+  GetPostByIdRoute,
+].map((route) => {
+  const { component: fn,  ...rest } = route;
   let component = null
   return {
```
</details>

TS can potentially tell _exactly_ which route we're dealing with and the specific values it has (e.g. `path: '/about'`  vs `path: string`) which would help connect the return types we're defining elsewhere

<details><summary>More screenshots</summary>
<img width="535" alt="Screen Shot 2021-08-23 at 8 53 36 PM" src="https://user-images.githubusercontent.com/57655/130539409-4bb3fce2-612e-4693-b1d6-6c6cd4d45f9f.png">
<img width="673" alt="Screen Shot 2021-08-23 at 8 55 01 PM" src="https://user-images.githubusercontent.com/57655/130539411-e6f8db9a-52bc-4ba1-bc29-1272239fa1a3.png">
<img width="614" alt="Screen Shot 2021-08-23 at 8 58 51 PM" src="https://user-images.githubusercontent.com/57655/130539414-568ff3c5-34ef-42f0-8256-8e9a1e8e57e5.png">
<img width="435" alt="Screen Shot 2021-08-23 at 9 02 02 PM" src="https://user-images.githubusercontent.com/57655/130539415-0497772b-6f9c-4347-b902-815c46de4ae5.png">
<img width="924" alt="Screen Shot 2021-08-23 at 9 02 11 PM" src="https://user-images.githubusercontent.com/57655/130539416-1b36f019-6276-4162-a68d-9f7f2d1ad2d6.png">
<img width="927" alt="Screen Shot 2021-08-23 at 9 02 34 PM" src="https://user-images.githubusercontent.com/57655/130539417-f64aba02-6b5f-403e-8152-e190d006667c.png">

</details>

**Comment by frandiox** [comment]
@jfsiii Thanks for the thorough explanation! Yeah, I totally agree with providing generics, but perhaps it would be better at the wrapper level?

```ts
declare module 'vitedge/define' {
  export const defineApiEndpoint: <T = any>(apiEndpoint: ApiEndpoint<T>) => ApiEndpoint<T>
  export const defineEdgeProps: <T = any>(edgeProps: EdgeProps<T>) => EdgeProps<T>
}
```

Or perhaps with 2 generics, one for the returned data type and another one for the `params`?

The hardest part I see is the "automatic connection" between frontend and backend. In  many cases the routes are created automatically using plugins (such as [`vite-plugin-pages`](https://github.com/hannoeru/vite-plugin-pages)), but even if we define routes manually with `const`, I'm not really sure how to take that to the backend 🤔 . Of course, any ideas or exploration in this area is very welcome!

As you say, worst case scenario the user can import the types manually. Maybe we could provide some aliases automatically so they can do `import type { MyProps } from '@props/my-page` easily?

**Comment by frandiox** [comment]
Some crazy idea that probably won't work with TS: create a virtual entry file that imports every prop handler type and re-exports the returned data type. Roughly this, not sure if this is even correct syntax:

```ts
import type { default as AboutHandler } from '..../functions/props/about'
export type AboutProps = ReturnType<AboutHandler['handler']>['data'] // This should also unwrap the Promise type
```

```ts
// About.jsx
import type { AboutProps } from 'virtual:vitedge-props'
```

Not even sure if this is desirable, just brainstorming 😅
